liferay/liferay-frontend-projects
# feat(liferay-cli): support for "liferay new" command to generate new projects

This is the PR for #556. See commits for more information.

This is fully functional for a Liferay Platform project based on [portal-7.4-ga1 platform](https://github.com/liferay/liferay-frontend-projects/tree/master/target-platforms/packages/portal-7.4-ga1), though it needs manual steps to be tested since we haven't released anything to npmjs.com yet. 

Once we all agree this is the way we want to go, I will migrate the rest of the projects from the old generator to the new CLI tool so that we can deprecate the old Yeoman generator and continue implementing new features in this CLI tool.

## How to test this locally

If you want to test this locally, you need a local npm registry (eg: [Verdaccio](https://verdaccio.org/)) or publishing the target platform as a pre-release in npmjs.com. Once that's done, you will need to use your local copy of `liferay-cli` to generate a project, and your local copy of the JS Toolkit v2 to build it.

See [this](https://github.com/liferay/liferay-frontend-projects/pull/562/files#diff-34eaddd24b1da53887651a803665379dff158ea8e900f4983fbfdfc5dc60f5e3R119) for more information and/or ask me for help if needed.

## How this looks when tested locally

In case you don't want to setup the local tests, I'm adding an explanation on how this works and pushing the results of the generator to the [fat-jar-sample](https://github.com/izaera/fat-jar-sample) repo, project [my-app](https://github.com/izaera/fat-jar-sample/tree/master/packages/my-app).

> :eyes: Note that you cannot use the `my-app` at `fat-jar-sample` to test the build locally because setting that up in a yarn workspace is even more complicated. So, use `my-app` only as a reference.

```sh
ivan@baco:~ $ liferay new my-app
👋 Welcome to Liferay Project Generator v1.0.0
🛈  Gathering info to create new project 'my-app'...
? What is the human readable description of your project? My App
? What will be your target platform? portal-7.4-ga1
? Do you want to add localization support? Yes
? Under which category should your widget be listed? category.sample
? Do you want to add configuration support?

  💡 Needs Liferay DXP/Portal CE 7.1 with JS Portlet Extender 1.1.0 or
     Liferay DXP/Portal CE 7.2+.

 Yes
🛈  Generating new project 'my-app'...
✅ Successfully generated new project 'my-app'

  You can now run the following commands to build your project:

      $ cd my-app ↩
      $ npm install ↩
      $ npm run build ↩

  This will create a my-app.jar file in your my-app/dist
  folder that you can deploy to your local Liferay server.


ivan@baco:~ $ cd my-app
ivan@baco:~/my-app $ yarn
yarn install v1.22.10
warning package.json: No license field
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
[4/4] Building fresh packages...
success Saved lockfile.
Done in 3.47s.
ivan@baco:~/my-app $ yarn build
yarn run v1.22.10
warning package.json: No license field
$ liferay build
Building project for target platform: @liferay/portal-7.4-ga1
ℹ️ Running babel on 2 files...
ℹ️ Running liferay-npm-bundler...
Copying package '/'...
Copied 1 packages
Applying rules to package '/'...
Applied rules to 1 packages
Transforming package '/'...
Babelifying 2 files in package '/'...
Transformed package '/'
Transformed 1 packages
Bundling took 265ms
The build has emitted some warning messages.
✔️ Project successfully built
Done in 1.71s.
ivan@baco:~/my-app $ cp dist/my-app-1.0.0.jar /opt/liferay/osgi/modules
ivan@baco:~/my-app $
```

And this is what you see in the Portal once you add the `my-app` portlet to a page:

![image](https://user-images.githubusercontent.com/3273630/120635379-b3622d80-c46c-11eb-96be-9bf749ad333b.png)

You can also access the two portlet configuration types: :one: in System Settings, :two: in portlet configuration dialog

I'm attaching the generated JAR file for inspection, too: [my-app-1.0.0.zip](https://github.com/liferay/liferay-frontend-projects/files/6591214/my-app-1.0.0.zip)
